### ISSUE TITLE
[Discussion] Recreate login

### ISSUE THREAD

**Issue by FabianTerhorst** [Grover-c13/PokeGOAPI-Java#660]
The current implementation is realy bad for catching errors. The exception is thrown on the constructor so that the object not even get created. We should think about something like api.login(provider, callback); The callback interface should have a onSuccess(profile, settings) and a onError(Throwable t, int errorCode); Or a LoginResult result = api.login(provider); for not async work.

**Comment by iGio90** [comment]
im + for a discussion about the login. I was about to create a new one so please, would you mind edit yours and add a todo list of what we have to do?
imo, this is what we have to do:
- fill the signature (done/almost done)

rework the login process:
- add the locale to the getPlayer request
- add all the missing login call (check dumps)
- a better error handling (as you said)

**Comment by iGio90** [comment]
i will add more when I start the rework to the login flow

**Comment by iGio90** [comment]
i'll also add
- improve the account verifier (https://github.com/Grover-c13/PokeGOAPI-Java/commit/289b7fcba28be01b985fd4dbbc8b831909977d9a) and make it fill the whole tutorial + setup the avatar + setup the nick

**Comment by FabianTerhorst** [comment]
we have to split the login out of the constructor.

**Comment by iGio90** [comment]
im agree, add it to the list

**Comment by iGio90** [comment]
if we split it out we would manage even better all the things i said in the list!

**Comment by FabianTerhorst** [comment]
so we shouldn´t create a boolean with first request.

**Comment by iGio90** [comment]
why not? all the things i set on the api (like the boolean and the timestamps) are managed by the signature class, it is safe to leave it there even with a splitted login!

**Comment by FabianTerhorst** [comment]
when we login there are other thinks we can check for this. For example player profile is null as well as settings.

**Comment by iGio90** [comment]
you are partially right. The player profile will be always null at the first request. The really first thing that the api do, is generate the signature for the first request. So, when you are going to generate the very first signature everything is still null! we need something to determine which one is the first generated signature because it send stripped data! You can rename the boolean from firstRequest to firstSignature if it would fit more, but we still need something that identify the first signature for the current api session

**Comment by iGio90** [comment]
or... yeah... as you said we can use if (getPlayerProfile == null) = first signature. But we have to verify this

**Comment by s7092910** [comment]
I would suggest moving from a Provider type system to a Provider and a Provided Credential.

Where the Provider is to create the Credential, which the Credential is the actual information needed to create the AuthToken and storing the information required to create that Credential. This allows better decoupling of the Provider and the API. Allowing more flexibility with the Provider.

**Comment by AbandonedCart** [comment]
@s7092910 You mean like https://github.com/Grover-c13/PokeGOAPI-Java/tree/Development/library/src/main/java/com/pokegoapi/auth (as in the current design)?

**Comment by s7092910** [comment]
@LoungeKatt The system you have now more like a Factory for AuthTokens. The issue is that the Factory itself requires knowledge of AuthTokens to create the AuthTokens. This requires a bit more work on the end when creating the Factory.

What a provider is, is sort of a Factory to a factory or a common factory interface which allows factories to be swappable. This means that the CredentialProviders are to create a Credential. Which then in turns creates a AuthToken. Frankly at most you only need 2 to 3 different Credentials, which the Credentials have have information on what provider created it, what values were required to create that Credential, when the Credential expires, etc. A factory is in theory not suppose to know everything about the object that it creates. The current system knows everything about the object.

On top of that, a Factory/Provider is suppose to be able to provide multiple different versions of an object based on inputs. So if I want to login 2 accounts, I require 2 Providers or if I supply one provider with the login details, and then supply it with the second login details, it will always return an authtoken based on the second login details.

What I'm suggesting is that You have 1 Provider and it can create as many Credentials as needed based on inputs which then creates the AuthTokens when needed. If you are using the same login type.

This allows more flexibility on your end, since a CredentialProvider creates a Credential, so if anyone needs to rewrite or extend or implement a CredentialProvider, they only need to create the Credential. Instead of having to give methods to know when to refresh the credentials or even how to fresh the credential. It makes it easier to maintain in the long run.

**Comment by AbandonedCart** [comment]
TL;DR
It sounds like an excessive conversion for multiple accounts based on more SOP, which has long been discarded due to the nature of the API itself.

What I mean is that you are more than welcome to submit a PR for it, but I don't see it being something urgent enough to justify making it a prerequisite to the current proposal.
